float64 xr
float64 yr
float64 thetar
float64 kr